Dockerfile

```diff
@@ -1,23 +1,16 @@
-# Use Python 3.11 base image
-FROM python:3.11-slim
+# Stage 1: Build dependencies and run tests
```

Comment:
The process has two stages:

1. **Build:** This involves tasks like installing dependencies and compiling code to create the needed artifacts. If files involved in building change, Docker will redo this stage.

2. **Run:** Here, artifacts from the build are copied into the final image, and the runtime environment is set up. If only application code or unrelated files change, Docker can reuse previous steps, making builds faster.